java/employee-scheduling/src/main/java/org/acme/employeescheduling/solver/EmployeeSchedulingConstraintProvider.java

```diff
@@ -54,18 +52,19 @@ Constraint requiredSkill(ConstraintFactory constraintFactory) {
     }
 
     Constraint noOverlappingShifts(ConstraintFactory constraintFactory) {
-        return constraintFactory.forEachUniquePair(Shift.class, Joiners.equal(Shift::getEmployee),
-                        Joiners.overlapping(Shift::getStart, Shift::getEnd))
+        return constraintFactory.forEachUniquePair(Shift.class, equal(Shift::getEmployee),
+                Joiners.overlapping(Shift::getStart, Shift::getEnd))
                 .penalize(HardSoftScore.ONE_HARD,
                         EmployeeSchedulingConstraintProvider::getMinuteOverlap)
                 .asConstraint("Overlapping shift");
     }
 
     Constraint atLeast10HoursBetweenTwoShifts(ConstraintFactory constraintFactory) {
         return constraintFactory.forEachUniquePair(Shift.class,
-                        Joiners.equal(Shift::getEmployee),
-                        Joiners.lessThanOrEqual(Shift::getEnd, Shift::getStart))
-                .filter((firstShift, secondShift) -> Duration.between(firstShift.getEnd(), secondShift.getStart()).toHours() < 10)
+                equal(Shift::getEmployee),
+                Joiners.lessThanOrEqual(Shift::getEnd, Shift::getStart))
```

Comment:
Are you sure about `forEachUniquePair` here? It is useful, but not in situations where the order of the two elements actually matters, such as here.

Consider shifts A, B. This will produce [A, B], but not [B, A]. This clashes with `A.end() < B.start()`; some of the pairs will not be selected and therefore not penalized.

A well-written test will catch this.

Reply:
Let me double-check it.

Reply:
Good catch! I'm able to reproduce it with the following test:

```
        constraintVerifier.verifyThat(EmployeeSchedulingConstraintProvider::atLeast10HoursBetweenTwoShifts)
                .given(employee1, employee2,
                        new Shift("1", DAY_END_TIME, DAY_START_TIME.plusDays(1), "Location", "Skill", employee1),
                        new Shift("2", DAY_START_TIME, DAY_END_TIME, "Location 2", "Skill", employee1))
                .penalizesBy(600);
```

Reply:
I double-checked the other constraints, and they look good to me. Also, the Python version is already using the `Join` operation.